### PR TITLE
[Merged by Bors] - Implement checkpoint sync

### LINKED ISSUE #1784: Remove canonical_head_block_root from PersistedBeaconChain
## Description

In #1639 the `canonical_head_block_root` field of `PersistedBeaconChain` was rendered obsolete by the use of fork choice to derive the head block on startup. We intended to remove it entirely when we did the breaking schema change for v0.3.0, but that PR (#1638) got closed and forgotten about :frowning_face: 

As a practice migration, I think we should remove the field entirely in a future release, i.e. automatically update the user's database from the old schema (with the block root) to the new (without) on startup.

### LINKED ISSUE #1891: Starting from a weak subjectivity point / arbitrary finalized epoch
## Description

It is desirable to avoid long syncs by simply supplying a trusted finalised state and starting lighthouse from this point. 

Some of the issues needed to support this are:
1. Handling block history - A LH node that starts at an arbitrary point does not have the block history and cannot respond correctly to it's peers requests who wish to sync from points before that arbitrary start. The spec indicates that we must maintain blocks from the most recent weak subjectivity point or in the case of long periods of non finality `min(finalized_epoch, weak_subjectivity_point)`. 
One solution is to attempt to download blocks backwards, however this runs into issues with block verification (in its current form) as we cannot verify blocks in a "backwards fashion". A somewhat cleaner solution would be to only allow LH to start from weak subjectivity points which allows us to maintain our sync logic and sync forward to the current head. The drawback here is if the weak subjectivity period is the order of months, we must spend the time to sync and verify all the blocks to the current head.

2. Handling peers in sync that only partially support the full history - We need to enhance a peer handling in sync to support peers with only partial views of the chain. I'll make a separate issue for this once we have decided on how peers will identify themselves as having a partial view of the block history.

### PR DESCRIPTION
**Pull request by michaelsproul**
## Issue Addressed

Closes #1891
Closes #1784

## Proposed Changes

Implement checkpoint sync for Lighthouse, enabling it to start from a weak subjectivity checkpoint.

## Additional Info

- [x] Return unavailable status for out-of-range blocks requested by peers (#2561)
- [x] Implement sync daemon for fetching historical blocks (#2561)
- [x] Verify chain hashes (either in `historical_blocks.rs` or the calling module)
- [x] Consistency check for initial block + state
- [x] Fetch the initial state and block from a beacon node HTTP endpoint
- [x] Don't crash fetching beacon states by slot from the API
- [x] Background service for state reconstruction, triggered by CLI flag or API call.

Considered out of scope for this PR:

- Drop the requirement to provide the `--checkpoint-block` (this would require some pretty heavy refactoring of block verification)

### CODE REVIEW

File: beacon_node/beacon_chain/src/historical_blocks.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/historical_blocks.rs]: I'm not sure what I was thinking here. The blocks should go in the hot database, not the cold, unless we change the schema.

**Comment by paulhauner** [comment]
I've tagged this for `v1.5.1`. That might be optimistic, but I just wanted to keep this on the radar. Feel free to downgrade :)

File: beacon_node/beacon_chain/src/historical_blocks.rs
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/historical_blocks.rs]: Fixed now, but needs a test

**Comment by michaelsproul** [comment]
I've made some tweaks to help with debugging:

* Update `lcli pretty-ssz` to print JSON, and support `SignedBeaconBlock`
* Add two new `/lighthouse/database` endpoints:
  * `/info`: info about the split slot, anchor point, and schema version
  * `/historical_blocks`:  POST endpoint for pushing historical blocks to a BN manually

My debugging flow (using scripts from [`eth2-scripts`](https://github.com/michaelsproul/eth2-scripts)) has been:

```
$ ssz_block.sh finalized > block.ssz
$ ssz_state.sh finalized > state.ssz
$ lighthouse bn --initial-state state.ssz --initial-block block.ssz --http
$ historical_sync.py
```

Now to fix the actual problems listed in the PR description :sweat_smile:

**Comment by michaelsproul** [comment]
There's an issue with the Altair beacon chain tests. Probably something benign, but I haven't had a chance to look into it yet.

**Comment by michaelsproul** [comment]
Also realised there are probably issues at skipped slots. If the initial block's slot doesn't match the state's I think we will run into problems. I'll look into the best way to handle this next week.

File: beacon_node/beacon_chain/src/historical_blocks.rs
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/historical_blocks.rs]: Tested now.

File: beacon_node/beacon_chain/src/builder.rs
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/builder.rs]: Incomplete sentence

**Comment by michaelsproul** [comment]
To send the `ResourceUnavailable` responses to `BlocksByRange` requests, I think we need to do something like:

Add a `ResourceUnavailable(String)` variant to `Response`:

https://github.com/sigp/lighthouse/blob/4283c1e98a680a910fc6a488826319b40886d2ee/beacon_node/eth2_libp2p/src/behaviour/mod.rs#L1168-L1175

Convert that to an `RPCCodedResponse::Error(RPCResponseErrorCode::ResourceUnavailable, message)`

https://github.com/sigp/lighthouse/blob/4283c1e98a680a910fc6a488826319b40886d2ee/beacon_node/eth2_libp2p/src/rpc/methods.rs#L261-L269

I think I could start hacking on this, but I might wait to see if you have any input @divagant-martian :blush: 

In the meantime I'll look at reconstructing the historical states in the DB, as this would really be a nice feature to have.

**Comment by michaelsproul** [comment]
Noting this here for later, something I encountered while working on https://github.com/sigp/lighthouse/issues/2526.

I reverted the head back to slot 1972768, but then sync was failing because it lacked the block at the previous slot 1972767

```
Aug 20 13:38:40.865 DEBG Sending Status Request                  fork_digest: [116, 101, 235, 11], finalized_epoch: 61647, finalized_root: 0xf896…fd84, head_slot: 1972768, head_root: 0xb4d6…c0a4, peer: 16Uiu2HAmTAJsRmASvFHWiH5vVUUQbMRnhTGT1Ez8GmTULGKC8Aco, service: router
Aug 20 13:38:40.944 DEBG Obtained peer's metadata                new_seq_no: 17, peer_id: 16Uiu2HAmTAJsRmASvFHWiH5vVUUQbMRnhTGT1Ez8GmTULGKC8Aco, service: libp2p
Aug 20 13:38:40.983 DEBG Received Status Response                fork_digest: [116, 101, 235, 11], finalized_epoch: 61794, finalized_root: 0x1af5…1684, head_slot: 1977492, head_root: 0xe9ed…c7b7, peer_id: 16Uiu2HAmTAJsRmASvFHWiH5vVUUQbMRnhTGT1Ez8GmTULGKC8Aco, service: router
Aug 20 13:38:40.983 DEBG Peer transitioned sync state            is_connected: true, their_finalized_epoch: 61794, their_head_slot: 1977492, out_finalized_epoch: 61647, our_head_slot: 1972768, new_state: Advanced, peer_id: 16Uiu2HAmTAJsRmASvFHWiH5vVUUQbMRnhTGT1Ez8GmTULGKC8Aco, service: sync
Aug 20 13:38:40.983 DEBG Finalization sync peer joined           peer_id: 16Uiu2HAmTAJsRmASvFHWiH5vVUUQbMRnhTGT1Ez8GmTULGKC8Aco, service: sync
Aug 20 13:38:40.984 DEBG New chain added to sync                 id: 10439739113780151862, from: 61647, to: 61794, end_root: 0x1af5…1684, current_target: 61647, batches: 0, peers: 1, validated_batches: 0, state: Stopped, sync_type: Finalized, peer_id: 16Uiu2HAmTAJsRmASvFHWiH5vVUUQbMRnhTGT1Ez8GmTULGKC8Aco, service: sync
Aug 20 13:38:40.984 DEBG Syncing new finalized chain             id: 10439739113780151862, from: 61647, to: 61794, end_root: 0x1af5…1684, current_target: 61647, batches: 0, peers: 1, validated_batches: 0, state: Stopped, service: sync
Aug 20 13:38:40.984 DEBG Requesting optimistic batch             start_slot: 1972769, end_slot: 1972832, downloaded: 0, processed: 0, state: Downloading(16Uiu2HAmTAJsRmASvFHWiH5vVUUQbMRnhTGT1Ez8GmTULGKC8Aco, 0 blocks, 1), epoch: 61649, chain: 10439739113780151862, service: sync
Aug 20 13:38:40.984 INFO Sync state updated                      new_state: Syncing Finalized Chain, old_state: Stalled, service: sync
Aug 20 13:38:41.364 DEBG Identified Peer                         protocols: ["/meshsub/1.1.0", "/meshsub/1.0.0", "/eth2/beacon_chain/req/status/1/ssz_snappy", "/eth2/beacon_chain/req/goodbye/1/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_range/2/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_range/1/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_root/2/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_root/1/ssz_snappy", "/eth2/beacon_chain/req/ping/1/ssz_snappy", "/eth2/beacon_chain/req/metadata/2/ssz_snappy", "/eth2/beacon_chain/req/metadata/1/ssz_snappy", "/ipfs/id/1.0.0", "/ipfs/id/push/1.0.0"], observed_address: "/ip4/115.64.39.178/tcp/53130", listening_ addresses: ["/ip4/202.87.162.38/tcp/9000", "/ip4/127.0.0.1/tcp/9000", "/ip4/192.168.20.2/tcp/9000"], agent_version: Lighthouse/v1.5.0-rc.1-c737983/x86_64-linux, protocol_version: eth2/1.0.0, peer: 16Uiu2HAmTAJsRmASvFHWiH5vVUUQbMRnhTGT1Ez8GmTULGKC8Aco, service: libp2p
Aug 20 13:38:41.398 DEBG Requesting new peer's metadata          peer_id: 16Uiu2HAmKxXQV9oMMY7ibRcPsiyRXvdLskbxFMobepyg428TwrEb, service: libp2p_rpc, service: libp2p
Aug 20 13:38:41.398 DEBG Connection established                  connections: 1, connection: Outgoing, peer_id: 16Uiu2HAmKxXQV9oMMY7ibRcPsiyRXvdLskbxFMobepyg428TwrEb, service: libp2p
Aug 20 13:38:41.398 DEBG Sending Status Request                  fork_digest: [116, 101, 235, 11], finalized_epoch: 61647, finalized_root: 0xf896…fd84, head_slot: 1972768, head_root: 0xb4d6…c0a4, peer: 16Uiu2HAmKxXQV9oMMY7ibRcPsiyRXvdLskbxFMobepyg428TwrEb, service: router
Aug 20 13:38:41.404 DEBG Requesting new peer's metadata          peer_id: 16Uiu2HAm9acuKiNKyCZcDRGXH33BRZrLiVxoYFGDgVqy8fCtGE8a, service: libp2p_rpc, service: libp2p
Aug 20 13:38:41.398 DEBG Connection established                  connections: 1, connection: Outgoing, peer_id: 16Uiu2HAmKxXQV9oMMY7ibRcPsiyRXvdLskbxFMobepyg428TwrEb, service: libp2p
Aug 20 13:38:41.398 DEBG Sending Status Request                  fork_digest: [116, 101, 235, 11], finalized_epoch: 61647, finalized_root: 0xf896…fd84, head_slot: 1972768, head_root: 0xb4d6…c0a4, peer: 16Uiu2HAmKxXQV9oMMY7ibRcPsiyRXvdLskbxFMobepyg428TwrEb, service: router
Aug 20 13:38:41.404 DEBG Requesting new peer's metadata          peer_id: 16Uiu2HAm9acuKiNKyCZcDRGXH33BRZrLiVxoYFGDgVqy8fCtGE8a, service: libp2p_rpc, service: libp2p
Aug 20 13:38:41.404 DEBG Connection established                  connections: 1, connection: Outgoing, peer_id: 16Uiu2HAm9acuKiNKyCZcDRGXH33BRZrLiVxoYFGDgVqy8fCtGE8a, service: libp2p
Aug 20 13:38:41.405 DEBG Sending Status Request                  fork_digest: [116, 101, 235, 11], finalized_epoch: 61647, finalized_root: 0xf896…fd84, head_slot: 1972768, head_root: 0xb4d6…c0a4, peer: 16Uiu2HAm9acuKiNKyCZcDRGXH33BRZrLiVxoYFGDgVqy8fCtGE8a, service: router
Aug 20 13:38:41.406 DEBG Peer disconnected                       peer_id: 16Uiu2HAm9acuKiNKyCZcDRGXH33BRZrLiVxoYFGDgVqy8fCtGE8a, service: libp2p
Aug 20 13:38:41.489 DEBG Requesting new peer's metadata          peer_id: 16Uiu2HAmVpu5NUWDsn4jnJWutHZbNcurL9rZvbkmZ1iREPG7EuEB, service: libp2p_rpc, service: libp2p
Aug 20 13:38:41.490 DEBG Connection established                  connections: 1, connection: Outgoing, peer_id: 16Uiu2HAmVpu5NUWDsn4jnJWutHZbNcurL9rZvbkmZ1iREPG7EuEB, service: libp2p
Aug 20 13:38:41.490 DEBG Sending Status Request                  fork_digest: [116, 101, 235, 11], finalized_epoch: 61647, finalized_root: 0xf896…fd84, head_slot: 1972768, head_root: 0xb4d6…c0a4, peer: 16Uiu2HAmVpu5NUWDsn4jnJWutHZbNcurL9rZvbkmZ1iREPG7EuEB, service: router
Aug 20 13:38:41.548 DEBG Requesting new peer's metadata          peer_id: 16Uiu2HAmCghybjva3o1BSrZJ4P1pwVpY8p917ZSEt9S5xmWwJkH7, service: libp2p_rpc, service: libp2p
Aug 20 13:38:41.548 DEBG Connection established                  connections: 1, connection: Outgoing, peer_id: 16Uiu2HAmCghybjva3o1BSrZJ4P1pwVpY8p917ZSEt9S5xmWwJkH7, service: libp2p
Aug 20 13:38:41.548 DEBG Sending Status Request                  fork_digest: [116, 101, 235, 11], finalized_epoch: 61647, finalized_root: 0xf896…fd84, head_slot: 1972768, head_root: 0xb4d6…c0a4, peer: 16Uiu2HAmCghybjva3o1BSrZJ4P1pwVpY8p917ZSEt9S5xmWwJkH7, service: router
Aug 20 13:38:41.551 DEBG Completed batch received                awaiting_batches: 0, blocks: 49, epoch: 61649, chain: 10439739113780151862, service: sync
Aug 20 13:38:41.551 DEBG Processing optimistic start             epoch: 61649, chain: 10439739113780151862, service: sync
Aug 20 13:38:41.573 DEBG Identified Peer                         protocols: ["/meshsub/1.1.0", "/meshsub/1.0.0", "/eth2/beacon_chain/req/status/1/ssz_snappy", "/eth2/beacon_chain/req/goodbye/1/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_range/2/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_range/1/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_root/2/ssz_snappy", "/eth2/beacon_chain/req/beacon_blocks_by_root/1/ssz_snappy", "/eth2/beacon_chain/req/ping/1/ssz_snappy", "/eth2/beacon_chain/req/metadata/2/ssz_snappy", "/eth2/beacon_chain/req/metadata/1/ssz_snappy", "/ipfs/id/1.0.0", "/ipfs/id/push/1.0.0"], observed_address: "/ip4/115.64.39.178/tcp/46574", listening_ addresses: ["/ip4/35.197.120.150/tcp/9852", "/ip4/172.17.0.1/tcp/9852", "/ip4/127.0.0.1/tcp/9852", "/ip4/10.138.0.13/tcp/9852"], agent_version: Lighthouse/v1.5.0-rc.1-c737983/x86_64-linux, protocol_version: eth2/1.0.0, peer: 16Uiu2HAmKxXQV9oMMY7ibRcPsiyRXvdLskbxFMobepyg428TwrEb, service: libp2p
Aug 20 13:38:41.574 DEBG Obtained peer's metadata                new_seq_no: 303, peer_id: 16Uiu2HAmKxXQV9oMMY7ibRcPsiyRXvdLskbxFMobepyg428TwrEb, service: libp2p
Aug 20 13:38:41.619 DEBG Requesting new peer's metadata          peer_id: 16Uiu2HAmMnq7YVtjc9BpKZjfaBLnRg8e4RtokHdFHhm7dTNZgcTX, service: libp2p_rpc, service: libp2p
Aug 20 13:38:41.619 DEBG Connection established                  connections: 1, connection: Outgoing, peer_id: 16Uiu2HAmMnq7YVtjc9BpKZjfaBLnRg8e4RtokHdFHhm7dTNZgcTX, service: libp2p
Aug 20 13:38:41.619 DEBG Sending Status Request                  fork_digest: [116, 101, 235, 11], finalized_epoch: 61647, finalized_root: 0xf896…fd84, head_slot: 1972768, head_root: 0xb4d6…c0a4, peer: 16Uiu2HAmMnq7YVtjc9BpKZjfaBLnRg8e4RtokHdFHhm7dTNZgcTX, service: router
Aug 20 13:38:41.634 WARN BlockProcessingFailure                  outcome: DBError(BlockNotFound(0x94558ab5bb50b71320d54cda47a45329d2a85d344bbab5fff4f4ab0c135f7fa6)), msg: unexpected condition in processing block.
Aug 20 13:38:41.634 DEBG Batch processing failed                 service: sync, error: Internal error whilst processing block: DBError(BlockNotFound(0x94558ab5bb50b71320d54cda47a45329d2a85d344bbab5fff4f4ab0c135f7fa6)), imported_blocks: 0, last_block_slot: 1972832, chain: 10439739113780151862, first_block_slot: 1972769, batch_epoch: 61649
```

Definitely something strange going on, as the log says `first_block_slot` is the slot after the head (1972769), which should be fine to apply. I guess it could be a fork, but I don't think so...

It seemed to fix itself once I back-filled a few blocks

**Comment by michaelsproul** [comment]
Dropping this from v1.5.1, as I think it will have to be in the release after

**Comment by michaelsproul** [comment]
Another sanity check to implement: check that the `genesis_validators_root` of the `--initial-state` matches the genesis state. I just spent too long accidentally trying to start a mainnet node with a Pyrmont state :man_facepalming:

**Comment by michaelsproul** [comment]
This is almost ready for review, just waiting on the merge of #2561.

There are now docs!

**Comment by michaelsproul** [comment]
This is ready for review! Batteries now included :tada:

File: beacon_node/beacon_chain/src/schema_change.rs
Comment (paulhauner) [inline beacon_node/beacon_chain/src/schema_change.rs:79]: Is this migration actually related to WSS? Or is it just #1784?

File: beacon_node/network/src/sync/backfill_sync/mod.rs
Comment (paulhauner) [inline beacon_node/network/src/sync/backfill_sync/mod.rs]: I thought I'd highlight this TODO in case it was forgotten about.

File: beacon_node/store/src/reconstruct.rs
Comment (paulhauner) [inline beacon_node/store/src/reconstruct.rs:103]: It's interesting that we don't check the state root unless there's a skip slot. This could mean, in a contrived scenario, after Altair we could break our base/phase0 impl without raising an error during a checkpoint sync.

In practice, skip slots will keep us honest. However there's something unsettling about relying on network failures to check integrity.

A simple way to address this might be to check the state root of the very last state that we import. This should be of negligible cost overall, but would give us some indication that we arrived at the same state.

Another option would be to ensure we do a state hash every `n` blocks, whether that's a skip slot one or an "artificial" forced one.

I'm still not sure this is worth doing, but I think it's at least worth raising.

File: book/src/checkpoint-sync.md
Comment (paulhauner) [inline book/src/checkpoint-sync.md]: Maybe we could add a note here to make it clear that you don't need those states for staking, just for historical API calls. That might avoid people opting into it "just to be safe".

Perhaps also that it will require significantly more space, even after the reconstruction process is complete.

File: consensus/state_processing/src/per_block_processing/block_signature_verifier.rs
Comment (paulhauner) [inline consensus/state_processing/src/per_block_processing/block_signature_verifier.rs]:
```suggestion
    /// Verify all the signatures that have been included in `self`, returning `true` if the
```

File: beacon_node/network/src/sync/manager.rs
Comment (paulhauner) [inline beacon_node/network/src/sync/manager.rs]:
```suggestion
    /// Updates the global sync state, optionally instigating or pausing a backfill sync as well as 
    /// logging any changes.
```

File: beacon_node/network/src/sync/backfill_sync/mod.rs
Comment (paulhauner) [inline beacon_node/network/src/sync/backfill_sync/mod.rs:725]:
```suggestion
                BatchState::Poisoned => {
                    crit!(self.log, "Poisoned batch");
                    unreachable!("Poisoned batch");
                }
```

**Review by paulhauner — CHANGES_REQUESTED** [review]
This is such an awesome feature, I had so much fun using it :tada: 

I've left a few rather minor suggestions and comments. I also have a few that don't relate to changes so I can't associate to a specific line:

- The `--shutdown-after-sync` flag will now shutdown after we reach head, but not before we finish back fill. I don't really know what the behaviour should be here, but at the very least we could add a message to the CLI help string flag indicating that this is the case.
- We also have the `--wss-checkpoint` flag on the BN. I think it should probably conflict with all the checkpoint-sync flags.

I also noticed we don't have the client-side of the new `/lighthouse` API endpoints. It would be fairly painless to add these and some smoke tests. I might pick this up when I get a chance and make a PR to this PR.

File: beacon_node/beacon_chain/src/schema_change.rs
Comment (michaelsproul) [reply beacon_node/beacon_chain/src/schema_change.rs:79]: The migration only makes changes related to #1784 but the schema version has to change because a WSS database adds the `AnchorInfo` which an earlier version would be oblivious to (to its peril).

File: beacon_node/network/src/sync/backfill_sync/mod.rs
Comment (michaelsproul) [reply beacon_node/network/src/sync/backfill_sync/mod.rs]: Idk what this is, I suspect it's been lurking in sync forever. @AgeManning would be the one to confirm.

File: consensus/state_processing/src/per_block_processing/block_signature_verifier.rs
Comment (michaelsproul) [reply consensus/state_processing/src/per_block_processing/block_signature_verifier.rs]: Fair point. I was trying to use "iff" to mean "if and only if", to clarify that the function returns true _only_ when the signatures are valid, i.e. it must return false when they are not. I will clarify the clarification :grin:

Comment (paulhauner) [reply consensus/state_processing/src/per_block_processing/block_signature_verifier.rs]: Oh sorry, I didn't realise `iff` was a thing!

File: beacon_node/network/src/sync/backfill_sync/mod.rs
Comment (AgeManning) [reply beacon_node/network/src/sync/backfill_sync/mod.rs]: Nah I added this to backfill sync (pretty sure). I'll make a decision on it.

File: beacon_node/network/src/sync/backfill_sync/mod.rs
Comment (AgeManning) [reply beacon_node/network/src/sync/backfill_sync/mod.rs:725]: This exists in normal sync also, can log a crit if we want, but it's unreachable ;)

Also, I think lighthouse will now log a crit on panic, so should already have a crit.

Comment (michaelsproul) [reply beacon_node/network/src/sync/backfill_sync/mod.rs:725]: Agree, QED :stuck_out_tongue:

File: book/src/checkpoint-sync.md
Comment (michaelsproul) [reply book/src/checkpoint-sync.md]: Done

File: beacon_node/store/src/reconstruct.rs
Comment (michaelsproul) [reply beacon_node/store/src/reconstruct.rs:103]: I've added a single state root hash at the end of sync, which required solving some issues with long gaps and the tree hash cache. Modified and tested in this commit: https://github.com/sigp/lighthouse/pull/2244/commits/30e2219947f15267571cab6ecd963780dd25823c

File: beacon_node/network/src/sync/backfill_sync/mod.rs
Comment (michaelsproul) [reply beacon_node/network/src/sync/backfill_sync/mod.rs]: Marking this as resolved, seeing as Age has fixed it.

**Comment by michaelsproul** [comment]
Thanks @paulhauner for the review, and thanks @AgeManning for fixing up that sync `TODO`. All comments have now been addressed, so I think we must be almost ready to merge :tada:

**Comment by michaelsproul** [comment]
Ah, I didn't address the flag-related suggestions from your main comment. Will do that now.

**Comment by michaelsproul** [comment]
Done.

Points to note:

* Deleted the `/lighthouse/backfill` API because its functionality was covered by `lighthouse/syncing`. Age was just using it during testing.
* The `lighthouse/database/historical_blocks` API is untested because it would be a pain to do a non-trivial test (requires a WSS node, and some blocks to POST). Seeing as it's for developers only and the docs state as much, I think this is an acceptable tradeoff.
* I've left `--wss-checkpoint` and `--checkpoint-*` as non-conflicting, because I think there are cases in which it could be useful to have both, e.g. if you start a checkpoint sync from several epochs back but want to verify that you reach some more recent checkpoint. This can be useful when you want quick access to some recent history, without having to backfill + reconstruct. I've updated the help text to point to `--checkpoint-sync-url`.

**Review by paulhauner — APPROVED** [review]
Awesome, I can't wait to see this in the hands of users!

You have my approval, merge at will capt'n :tada:

**Comment by michaelsproul** [comment]
:open_mouth: :open_mouth: :open_mouth:

bors r+

**Pull request closed without merging**
